<h1 align=center><font size = 5>Graded Assignment Project - Regression Model in Keras</font></h1>


## Introduction


In this course project, I will build a regression model using the deep learning Keras library, and then experiment with increasing the number of training epochs and changing number of hidden layers to find out how changing these parameters impacts the performance of the model.


<h4>Objective for this Assignment<h4>    
<h5> A. Build a baseline model </h5>
<h5> B. Normalize the data </h5>
<h5> C. Increase the number of epochs </h5>
<h5> D. Increase the number of hidden layers </h5>   


## Table of Contents

<div class="alert alert-block alert-info" style="margin-top: 20px">

<font size = 3>
    
<a href="#item31">A. Build a baseline model</a>  
<a href="#item32">B. Normalize the data</a>  
<a href="#item33">C. Increase the number of epochs </a>  
<a href="#item34">D. Increase the number of hidden layers </a>  

</font>
</div>


<a id="item31"></a>


## A. Build a baseline model

In this first step, we will use the Keras library to build a neural network with the following:
- One hidden layer of 10 nodes, and a ReLU activation function
- Use the adam optimizer and the mean squared error  as the loss function.


#### Import the Pandas and the Numpy libraries.


In [2]:
import pandas as pd
import numpy as np

The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. 
    
The data has been downloaded from: https://cocl.us/concrete_data.
    
The predictors in the data of concrete strength include include:
1. Cement
2. Blast Furnace Slag
3. Fly Ash
4. Water
5. Superplasticizer
6. Coarse Aggregate
7. Fine Aggregate


#### Read the data into a Pandas dataframe.


In [3]:
concrete_data = pd.read_csv('https://cocl.us/concrete_data')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

As can be seen from the data, the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. This concrete mix is 28 days old and has a compressive strength of 79.99 MPa. 


#### Split data into predictors and target


The target variable in this case is the concrete sample strength. Therefore, all the columns will be considered as predictors.


In [4]:
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

<a id="item2"></a>


#### Sanity check of the predictors and the target dataframes.


In [5]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [6]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

#### Save the number of predictors to _n_cols_ since we will need this number later on.


In [7]:
n_cols = predictors.shape[1] # number of predictors

#### Import the Keras library


In [8]:
import keras

Using TensorFlow backend.
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:519: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:520: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:521: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/home/jupyterlab/con

#### Import the rest of the packages from the Keras library that we will need to build the regressoin model.


In [9]:
from keras.models import Sequential
from keras.layers import Dense

### Build a Neural Network


Define a function that defines the regression model that can be conveniently called it to create our model. Use the Keras library to build a neural network with the following:

- One hidden layer of 10 nodes, and a ReLU activation function
- Use the adam optimizer and the mean squared error  as the loss function.


In [10]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

<a id="item4"></a>


### Train and Test the Neural Network

#### Create the model.


In [12]:
# build the model
model = regression_model()

#### 1. Randomly split the data into a training and test sets by holding 30% of the data for testing, using the train_test_split helper function from scikit-learn


In [13]:
from sklearn.model_selection import train_test_split

In [14]:
Xtrain, Xtest, ytrain, ytest = train_test_split(predictors, target, test_size=0.3)

#### 2. Train the model on the training data using 50 epochs.

In [16]:
model.fit(Xtrain, ytrain, epochs=50, verbose=1)

Epoch 1/50
721/721 [==============================] - 0s 389us/step - loss: 120.0485
Epoch 2/50
721/721 [==============================] - 0s 358us/step - loss: 121.8371
Epoch 3/50
721/721 [==============================] - 0s 305us/step - loss: 119.0435
Epoch 4/50
721/721 [==============================] - 0s 308us/step - loss: 118.8093
Epoch 5/50
721/721 [==============================] - 0s 332us/step - loss: 118.0583
Epoch 6/50
721/721 [==============================] - 0s 327us/step - loss: 117.4470
Epoch 7/50
721/721 [==============================] - 2s 2ms/step - loss: 115.9311
Epoch 8/50
721/721 [==============================] - 0s 366us/step - loss: 115.0056
Epoch 9/50
721/721 [==============================] - 0s 328us/step - loss: 114.5506
Epoch 10/50
721/721 [==============================] - 0s 663us/step - loss: 114.2920
Epoch 11/50
721/721 [==============================] - 0s 583us/step - loss: 113.4235
Epoch 12/50
721/721 [==============================] - 0s 579us/s

#### 3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength, using the mean_squared_error function from scikit-learn.

In [17]:
from sklearn.metrics import mean_squared_error

In [ ]:
MSE = mean_squared_error (ytest, model.predict(Xtest))
print ("MSE between predicted and actual concrete strength: ", MSE)

#### 4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors

In [19]:
# build the model
model = regression_model()

In [ ]:
MSE = np.empty(50)
for i in range(50):
    Xtrain, Xtest, ytrain, ytest = train_test_split(predictors, target, test_size=0.3)
    model.fit(Xtrain, ytrain, epochs=50, verbose=0)
    MSE[i] = mean_squared_error (ytest, model.predict(Xtest))
    
print ("Part A: List of Mean Squared Errors: ", MSE)

#### 5. Report the mean and the standard deviation of the mean squared errors.

In [ ]:
MSE_mean = MSE.mean()
MSE_std = MSE.std()

print ("Part A: Mean of the Mean Squared Errors: ", MSE_mean)
print ("Part A: Standard Deviation of the Mean Squared Errors: ", MSE_std)

## B. Normalize the data

Repeating Part A using a normalized version of the data

Q) How does the mean of the mean squared errors compare to that from Step A?


In [ ]:
#### Normalize the data by substracting the mean from the individual predictors and dividing by the standard deviation.

In [ ]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

In [ ]:
n_cols = predictors_norm.shape[1] # number of predictors

In [ ]:
# build the model for Part B using normalized data
model = regression_model()

In [ ]:
# repeat steps in Part A to create a list of 50 mean squared errors (using normalized data model)

In [ ]:
MSE = np.empty(50)
for i in range(50):
    Xtrain, Xtest, ytrain, ytest = train_test_split(predictors, target, test_size=0.3)
    model.fit(Xtrain, ytrain, epochs=50, verbose=0)
    MSE[i] = mean_squared_error (ytest, model.predict(Xtest))
    
print ("Part B: List of Mean Squared Errors: ", MSE)

In [ ]:
# calculating mean and standard deviation for Part B

In [ ]:
MSE_mean = MSE.mean()
MSE_std = MSE.std()

print ("Part B: Mean of the Mean Squared Errors: ", MSE_mean)
print ("Part B: Standard Deviation of the Mean Squared Errors: ", MSE_std)

In [ ]:
## C. Increase the number of epochs

Repeat Part B using 100 epochs this time for training.

Q) How does the mean of the mean squared errors compare to that from Step B?

In [ ]:
# repeating steps in Part B using 100 epochs for training

In [ ]:
# build the model for Part C
model = regression_model()

In [ ]:
MSE = np.empty(50)
for i in range(50):
    Xtrain, Xtest, ytrain, ytest = train_test_split(predictors, target, test_size=0.3)
    model.fit(Xtrain, ytrain, epochs=100, verbose=0)
    MSE[i] = mean_squared_error (ytest, model.predict(Xtest))
    
print ("Part C: List of Mean Squared Errors: ", MSE)

In [ ]:
# calculating mean and standard deviation for Part C

In [ ]:
MSE_mean = MSE.mean()
MSE_std = MSE.std()

print ("Part C: Mean of the Mean Squared Errors: ", MSE_mean)
print ("Part C: Standard Deviation of the Mean Squared Errors: ", MSE_std)

In [ ]:
## D. Increase the number of hidden layers

Repeat part B using a neural network with the following:

- Three hidden layers, each of 10 nodes and ReLU activation function.

Q) How does the mean of the mean squared errors compare to that from Step B?

In [ ]:
# define regression model with 3 hidden layers, each of 10 nodes and ReLU activation functions
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

In [ ]:
# build the model for Part D
model = regression_model()

In [ ]:
# repeating steps in Part B using 50 epochs for training the neural network of Part D

In [ ]:
MSE = np.empty(50)
for i in range(50):
    Xtrain, Xtest, ytrain, ytest = train_test_split(predictors, target, test_size=0.3)
    model.fit(Xtrain, ytrain, epochs=50, verbose=0)
    MSE[i] = mean_squared_error (ytest, model.predict(Xtest))
    
print ("Part D: List of Mean Squared Errors: ", MSE)

In [ ]:
# calculating mean and standard deviation for Part D

In [1]:
MSE_mean = MSE.mean()
MSE_std = MSE.std()

print ("Part D: Mean of the Mean Squared Errors: ", MSE_mean)
print ("Part D: Standard Deviation of the Mean Squared Errors: ", MSE_std)

NameError: name 'MSE' is not defined